client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1695 @@
+		let _ = self
+			.events_tx
+			.unbounded_send(Event::IncomingConnection(peer_id, incoming_index));
+	}
+
+	/// Notify that connection was dropped (either refused or disconnected).
+	pub fn dropped(&self, peer_id: PeerId) {
+		let _ = self.events_tx.unbounded_send(Event::Dropped(peer_id));
+	}
+}
+
+/// Direction of a connection
+#[derive(Clone, Copy, Debug)]
+enum Direction {
+	Inbound,
+	Outbound,
+}
+
+/// Status of a connection with a peer.
+#[derive(Clone, Debug)]
+enum PeerState {
+	/// We are connected to the peer.
+	Connected(Direction),
+	/// We are not connected.
+	NotConnected,
+}
+
+impl PeerState {
+	/// Returns true if we are connected with the node.
+	fn is_connected(&self) -> bool {
+		matches!(self, PeerState::Connected(_))
+	}
+}
+
+impl Default for PeerState {
+	fn default() -> PeerState {
+		PeerState::NotConnected
+	}
+}
+
+/// Side of [`ProtocolHandle`] responsible for all the logic. Currently all instances are
+/// owned by [`crate::Peerset`], but they should eventually be moved to corresponding protocols.
+#[derive(Debug)]
+pub struct ProtocolController {
+	/// Set id to use when sending connect/drop requests to `Notifications`.
+	// Will likely be replaced by `ProtocolName` in the future.
+	set_id: SetId,
+	/// Receiver for outer API messages from [`ProtocolHandle`].
+	actions_rx: TracingUnboundedReceiver<Action>,
+	/// Receiver for connection events from `Notifications` sent via [`ProtocolHandle`].
+	events_rx: TracingUnboundedReceiver<Event>,
+	/// Number of occupied slots for incoming connections (not counting reserved nodes).
+	num_in: u32,
+	/// Number of occupied slots for outgoing connections (not counting reserved nodes).
+	num_out: u32,
+	/// Maximum number of slots for incoming connections (not counting reserved nodes).
+	max_in: u32,
+	/// Maximum number of slots for outgoing connections (not counting reserved nodes).
+	max_out: u32,
+	/// Connected regular nodes.
+	nodes: HashMap<PeerId, Direction>,
+	/// Reserved nodes. Should be always connected and do not occupy peer slots.
+	reserved_nodes: HashMap<PeerId, PeerState>,
+	/// Connect only to reserved nodes.
+	reserved_only: bool,
+	/// Next time to allocate slots. This is done once per second.
+	next_periodic_alloc_slots: Instant,
+	/// Outgoing channel for messages to `Notifications`.
+	to_notifications: TracingUnboundedSender<Message>,
+	/// `PeerStore` handle for checking peer reputation values and getting connection candidates
+	/// with highest reputation.
+	peer_store: Box<dyn PeerReputationProvider>,
+}
+
+impl ProtocolController {
+	/// Construct new [`ProtocolController`].
+	pub fn new(
+		set_id: SetId,
+		config: SetConfig,
+		to_notifications: TracingUnboundedSender<Message>,
+		peer_store: Box<dyn PeerReputationProvider>,
+	) -> (ProtocolHandle, ProtocolController) {
+		let (actions_tx, actions_rx) = tracing_unbounded("mpsc_api_protocol", 10_000);
+		let (events_tx, events_rx) = tracing_unbounded("mpsc_notifications_protocol", 10_000);
```

Comment:
I don't see a reason why these would have to be unbounded. Maybe right now it's too much work get rid of them but if the network is opening more substreams faster than the protocol can handle them, `Notifications` should drop the substreams.

Reply:
Created an issue https://github.com/paritytech/substrate/issues/14173 for this.